### ISSUE TITLE
POD + SA restores fail due to invalid token

### ISSUE THREAD

**Issue by james-powis** [heptio/ark#840]
**What steps did you take and what happened:**

Prerequisits: 
- Persistent volumes - I use https://github.com/vmware/vsphere-storage-for-kubernetes
- Restic configured ark
- Helm
- Ingress Controller (I use haproxy)

Steps to reproduce working example:
1. Create values-works.yaml:
```yaml
Master:
  ServiceType: ClusterIP
  HostName: jenkins.example.com
  PodAnnotations:
    backup.ark.heptio.com/backup-volumes: jenkins-home
Persistence:
  Enabled: true
  AccessMode: ReadWriteOnce
  Size: 8Gi
```
2. Deploy helm chart and create backup.
```
helm upgrade --install test-jenkins stable/jenkins -f values-works.yaml
# Monitor for everything being complete
ark create backup test-jenkins-works
# Monitor for backup completion.
```
3. Delete the helm chart and restore from backup
```
helm delete --purge test-jenkins
# Monitor for all objects being removed
ark create restore test-jenkins-works --from-backup test-jenkins-works
# Monitor the restore, it will complete successfully.
# Cleanup for next example:
helm delete --purge test-jenkins
```

Steps to create the issue which breaks the restore (all we are adding is the service account):
1. Create values-broken.yaml
```yaml
Master:
  ServiceType: ClusterIP
  HostName: jenkins.example.com
  PodAnnotations:
    backup.ark.heptio.com/backup-volumes: jenkins-home
Persistence:
  Enabled: true
  AccessMode: ReadWriteOnce
  Size: 8Gi
rbac:
  install: true
  serviceAccountName: jenkins
  apiVersion: v1
  roleRef: cluster-admin
  roleBindingKind: ClusterRoleBinding
```
2. Deploy helm chart and create backup.
```
helm upgrade --install test-jenkins stable/jenkins -f values-broken.yaml
# Monitor for everything being complete
ark create backup test-jenkins-broken
# Monitor for backup completion.
```
3. Delete the helm chart and restore from backup
```
helm delete --purge test-jenkins
# Monitor for all objects being removed
ark create restore test-jenkins-broken --from-backup test-jenkins-broken
# Monitor the restore, it will not complete
```

Cleanup... (note: this might not be the best way to clean everything up, but it worked for me.)

1. Delete the helm chart `helm delete --purge test-jenkins`
2. Remove the restore which is hung in progress: `ark delete restore test-jenkins`
3. For good measure delete the ark pod because it is hung, it will recreate automatically `kubectl -n heptio-ark delete pod ark-asdf1234`

**What did you expect to happen:**

As demonstrated from `kubectl get pod test-jenkins-asdf1234 -o yaml` we can see that restic injects an int controller into the pod, in the absence of the RBAC configurations, this is successful, but in the presence of RBAC this fails. The delta of these 2 different configs is here: https://gist.github.com/james-powis/b289b5a9467350e5dc7a95b12ec6388e

And the hung restore state can be observed here https://gist.github.com/james-powis/0b5854fb89784e4ce5e1b6aab9458f61

**Environment:**

- Ark version (use `ark version`):  
```
Version: v0.9.3
Git commit: v0.9.3
Git tree state: clean
```
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:17:28Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version: 
```
kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:14:39Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: vcenter/vsphere
- OS (e.g. from `/etc/os-release`):
```
# cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Comment by james-powis** [comment]
@jrnt30 I am having difficulties getting the restore to fail today using minimal jenkins installs, so here is a complete example we are using. and the requested get pod -o json as requested. 

https://gist.github.com/james-powis/bf82bee13d09f4bdf833a87c63e68552

**Comment by james-powis** [comment]
Also here is the computed k8s manifest from helm: https://gist.github.com/james-powis/223511e51315d1511855c934192e441c

**Comment by james-powis** [comment]
Just realized in my examples provided in the initial issue I was missing the heptio backup annotations. This is why I was having difficulties replicating this morning... I now have a much more simple breakdown which causes the issue I am going to edit the initial report to remove a lot of the confusion.

**Comment by james-powis** [comment]
@jrnt30 I suspect you are right in your assumption. Upon further investigation I have cleaned up this issue and improved the narrative so it reads easier for future generations. The initial report now contains links to all the relevant information you requested, as well as a diff between a working example and a non working example. The main difference Service Account and RBAC.

**Comment by james-powis** [comment]
@ncdc Correct the pod fails to start with the Ark added init.

**Comment by ncdc** [comment]
@james-powis it's not specifically the ark-added init container that's the problem, I don't think. It's because it's trying to mount a secret that doesn't exist in the set of restored resources. Can you provide the yaml for the pod, serviceaccount(s), secret(s) (not the secret data, just the names and labels) both in the original form, and post-restore?

**Comment by james-powis** [comment]
@ncdc the tokens are dynamically created when a ServiceAccount is created (or in this case restored). This causes the tokens ID to drift from that ark recorded in the backup. Thus we end up with volumes for the old and new secret in the Init container copy-default-config as well as a merging of old and new volumes in the actual pod container.

this can be observed here:

https://gist.github.com/james-powis/0b5854fb89784e4ce5e1b6aab9458f61
With the yaml here (note it is a diff of works and does not work): 
https://gist.github.com/james-powis/b289b5a9467350e5dc7a95b12ec6388e

**Comment by james-powis** [comment]
OK got this down to the absolute minimal POI.... You should be able to replicate locally without too much effort.

```root@k8sm01:~/helm/jenkins/ark-testing# cat example.yaml 
apiVersion: v1
kind: ServiceAccount
metadata:
  name: my-sa
  namespace: default
---
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    ports:
    - containerPort: 80
  serviceAccountName: my-sa

root@k8sm01:~/helm/jenkins/ark-testing# kubectl apply -f example.yaml 
serviceaccount/my-sa created
pod/nginx created
root@k8sm01:~/helm/jenkins/ark-testing# ark create backup example
Backup request "example" submitted successfully.
Run `ark backup describe example` for more details.

root@k8sm01:~/helm/jenkins/ark-testing# kubectl delete -f example.yaml 
serviceaccount "my-sa" deleted
pod "nginx" deleted
root@k8sm01:~/helm/jenkins/ark-testing# ark restore create example --from-backup example.yaml 
An error occurred: backups.ark.heptio.com "example.yaml" not found
root@k8sm01:~/helm/jenkins/ark-testing# ark restore create example --from-backup example
Restore request "example" submitted successfully.
Run `ark restore describe example` for more details.

root@k8sm01:~/helm/jenkins/ark-testing# kubectl describe pod nginx
Name:               nginx
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               k8sc03/172.30.122.45
Start Time:         Wed, 12 Sep 2018 14:51:07 -0600
Labels:             ark-restore=example
Annotations:        kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"nginx","namespace":"default"},"spec":{"containers":[{"image":"nginx:1.7.9","name":...
Status:             Pending
IP:                 
Containers:
  nginx:
    Container ID:   
    Image:          nginx:1.7.9
    Image ID:       
    Port:           80/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from my-sa-token-wldjj (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  my-sa-token-wldjj:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  my-sa-token-wldjj
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason       Age                From               Message
  ----     ------       ----               ----               -------
  Normal   Scheduled    44s                default-scheduler  Successfully assigned default/nginx to k8sc03
  Warning  FailedMount  11s (x7 over 43s)  kubelet, k8sc03    MountVolume.SetUp failed for volume "my-sa-token-wldjj" : secrets "my-sa-token-wldjj" not found
```
Now we see the old and the new token in the sa config.
```
root@k8sm01:~/helm/jenkins/ark-testing# kubectl get sa my-sa -o yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"name":"my-sa","namespace":"default"}}
  creationTimestamp: 2018-09-12T20:50:48Z
  labels:
    ark-restore: example
  name: my-sa
  namespace: default
  resourceVersion: "2756702"
  selfLink: /api/v1/namespaces/default/serviceaccounts/my-sa
  uid: 86d9afc3-b6cd-11e8-a6ca-005056aea577
secrets:
- name: my-sa-token-wldjj
- name: my-sa-token-tcn94
```

**Comment by james-powis** [comment]
OK I think i know enough to try my hand at submitting a PR for this... /holdmybeer

**Comment by james-powis** [comment]
@ncdc if you have a moment can you give pr #843 a once over.

**Comment by nrb** [comment]
@james-powis Thanks for the PR. All of us are in meetings this week, so full reviews will probably be a little delayed.

**Comment by ncdc** [comment]
Thanks! Our team is in all day meetings this week but hopefully we can
review in between sessions.

**Comment by james-powis** [comment]
*JINX* HAHA Ok no worries, We are anxious to have working backup and restore as most of or k8s cluster is going to rely heavily upon RBAC SA Tokens, so I will do my best to resolve any issues you may find with the PR relatively quickly.

**Comment by skriss** [comment]
Fixed by #843
